Fleet Sampling docs

can-memfault opened the pull request.

HBehrens commented:
Thanks for this PR, @can-memfault – will probably not get to a review today but added it to my list.

HBehrens commented on sidebars.js:
```diff
@@ -24,6 +24,7 @@ module.exports = {
                 "platform/charts",
                 "platform/alerts",
                 "platform/notifications",
+                "platform/fleet-sampling",
```
just a thought – nothing to act on here: At some point we need to rethink our information architecture… For now, I think this is the right place.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+---
+
+As your fleet grows, it becomes more costly to send, process and store the data
+from each one of your devices. The fleet sampling feature will help keeping the
```
```suggestion
from each one of your devices. Fleet Sampling will help keeping the
```

Here and in other places, I think we should treat this term as a concepts similar to "Cohort" and write it with uppercase letters. The article does this already on most places but not everywhere.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+As your fleet grows, it becomes more costly to send, process and store the data
+from each one of your devices. The fleet sampling feature will help keeping the
+costs low by collecting diagnostics and performance data only from a subset of
+your fleet but at the same time provide enough insights to understand issues
+happening across your fleet.
```
```suggestion
costs low by collecting diagnostics and performance data only from a smaller, yet statistically significant, subset of
your fleet. At large fleet sizes, this smaller "sampled" population of devices provide sufficiently representative data to still provide all relevant insights and to understand issues
as they occur across your fleet.
```

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+:::note
+
+Fleet Sampling is only supported on the
+[Bort SDK](https://github.com/memfault/bort), starting with
+[version 4.2.0](https://github.com/memfault/bort/blob/master/CHANGELOG.md#v420---september-14th-2022).
```
This is not exactly true. On all platforms, data from devices will be dropped upon ingestion if subject to sampling. What we do on Bort is device-side logic to prevent data from being sent but preserved in the holding area for potential later use (especially helpful for logs - which only exist on Android anyway).

I'd scratch this section here and maybe call it out further down.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+:::note
+
+Memfault strongly recommends usage of
+[Normalized Charts](/docs/platform/charts/#chart-normalization) on Dashboard and
+Metric Charts with Fleet Sampling to understand trends across your fleet
+correctly since the population of devices reporting issues from the field can
+change over time with the updates to diagnostic or performance reporting aspects
+of the devices.
+
+e.g. When every device of a million-device fleet crashes once a day this would
+normally be 1M crashes per day. However, when only 5% of the devices in the
+fleet are set to sent crash reports to Memfault, this might be reduced to
+50K/day, which would suggest a drop in crash counts, so to counteract this
+skewing of data, chart normalization will instead report 1K₁ₖ (one thousand
+crashes per thousand devices) for both scenarios.
```
How about "enables Fleet Sampling by default".

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+:::
+
+## Sampling Aspects
```
```suggestion
## Fleet Sampling Aspects
```

SEO

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+## Sampling Aspects
+
+Each device in the fleet has 3 aspects that represent different core features
+offered by Memfault. See the differences between the aspects and their
```
```suggestion
Each device in the fleet can be configured by turning on or off specific Fleet Sampling Aspects. See the differences between the aspects and their
```

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+| Debugging           | No crash collection                                | Crash report collection (via Bug Reports or Caliper), reboot reasons                                                                                                                                                                                                                               |
+| Logging             | No logs, except if they are part of a crash report | Log collection and upload in the case of an overlap with a trace event (see details [here](/docs/android/android-logging/)). <br />Logs that have been already collected on the device and not yet sent will also be uploaded as soon as the device receives the logging resolution change to `On` |
+
+Your devices won’t be “in the dark” even if all three aspect resolutions are set
```
The term "resolution" here, throughout the article, and frankly in our UI, is a bit confusing now that reduced this to On/Off.

Should we wat least try to avoid it in the docs? Where "set resolution for an aspect" would become something along the lines of "to turn on/off an aspect" ?

HBehrens reviewed (commented):
Only got as far as "Managing aspect resolutions of your devices" today. Will continue later.

Left some inline comments and suggestions. Also, I believe the term "Fleet Sampling" should be added to the glossary.

tyhoff commented:
Square is asking for hte docs for fleet sampling - heads up. Even if it's in preview, can send something early over and leave it off the sidebar for a little.

tyhoff commented:
What is the status here?

can-memfault commented:
@tyhoff sorry for the late response, Heiko and I will be reviewing the changes. Is there another customer waiting for it? I am only aware of Square

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+![fleet_sampling_quota_page.png](/img/docs/platform/fleet_sampling_quota_page.png)
+_Figure 4: Quotas and usage information_
+
+### Setting aspect resolution(s) of multiple devices
```
is this the right level of headlines? Making it level 3 makes it appear as if it was details about quotas.

![Screen Shot 2022-12-14 at 11 02 52](https://user-images.githubusercontent.com/486904/207565748-b375f3ea-18ea-4deb-8051-da216cac1719.png)

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+Memfault’s
+[Device Search](https://app.memfault.com/organizations/-/projects/-/devices)
+allows one to precisely describe a population of devices before assigning their
```
```suggestion
allows you to precisely describe a population of devices before assigning their
```

This is the only time in the doc you're using 3rd person instead of 2nd?

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+Memfault’s
+[Device Search](https://app.memfault.com/organizations/-/projects/-/devices)
+allows one to precisely describe a population of devices before assigning their
+respective sampling resolutions. Using the search parameters, specific parts of
```
```suggestion
respective sampling resolutions. Using the search parameters, specific populations of
```

Trying to gradually introducing the term "device population" across the product

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+the fleet where the most visibility is needed (i.e. devices experiencing fast
+battery discharges/connectivity issues in the past or devices in a specific
+cohort) can be defined and their sampling resolutions can be updated
+accordingly. Having such a visibility is also important before rolling out new
```
```suggestion
all at once in bulk. Having such a visibility is also important before rolling out new
```

Wanted to emphasize that this is for many and also to introduce the term "bulk" as this is being used in other places in the doc but not here where we explain it.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+In the event of hitting [quota limits](#quotas) when assigning sampling
+resolutions in bulk, a warning message will be presented to the user (see the
+screenshot below). Memfault strongly recommends freeing up the required quota
+for the assignment before continuing with the assignment. **Add why there may be
+surprises if over quota assignment is done for multiple aspects** This can be
```
```suggestion
screenshot below). You should free up the required quota
for the assignment before continuing with the assignment. This can be
```

The "strongly recommended" sounds weird. I also removed the note "add why there may be surprises" – do we really need that or can we explain this in a comprehensible way?

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+surprises if over quota assignment is done for multiple aspects** This can be
+done via:
+
+1. Updating your plan by emailing sales@memfault.com
```
```suggestion
1. Expanding your plan by emailing sales@memfault.com
```

Just curious: Do we have precedence for this in our docs? "Pay here"? I feel we should have a dedicated page for this at some point

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+done via:
+
+1. Updating your plan by emailing sales@memfault.com
+2. Turning the respective resolutions of some other devices to `Off`
```
```suggestion
2. Turning the respective resolutions of some other devices to `Off` by using the same mechanism as a prior step.
```

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+hitting the quota limit for Logging_
+
+Another option for predictably assigning sampling resolutions is to “limit” the
+number of devices that’ll be affected from the bulk operation: It limits the
```
```suggestion
number of devices that will be affected from the bulk operation: It limits the
```

nit: I believe [we once discussed](https://www.notion.so/memfault/How-to-Write-Copy-a5a89c8de6934ba4ad5d8923e1df8c40#d9e0f7893ebb40e1bf15a57450a9073f) "no abbreviations" in our style guide (there's a second place in this doc). I am sure this is very inconsistent throughout our docs in general, though.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+assignment to be applied on `the first N devices` that match with the search
+criteria and the sorting order. In the example below, the quota for logging is
+limited to 10 but 80 devices match the search query `software_version = 1.0.0`:
+Using the limit option, the change will be applied on the last seen 5 devices
```
This is so cool!

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+As this operation can take a long time, the result will be communicated via
+email to the user initiated the action. The email contains a summary of the
+changes and how many devices are affected by the change:
+
+![fleet_sampling_email.png](/img/docs/platform/fleet_sampling_email.png)
```
```suggestion
changes and how many devices are affected by the change.
```

I would not show a screenshot of the email excerpt. It does add little value and keeping it up-to-date will be very difficult.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+new devices may have contacted Memfault and have a more recent last seen
+information in the meantime.
+
+If devices to be updated with new sampling resolutions need to be precisely
+selected, checkboxes can be used and in this case only the selected devices will
+be updated.
```
```suggestion
If devices to be updated with new sampling resolutions need to be precisely
selected, please select them explicitly via the checkboxes in the search result before performing the action.
```

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+
+### Default fleet sampling configuration of new devices
+
+Memfault will assign best possible resolutions to the newly enrolled devices, as
```
```suggestion
Memfault will automatically assign resolutions to the newly enrolled devices, as
```

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+[Setting aspect resolution of multiple devices](#setting-aspect-resolutions-of-multiple-devices)
+section to free up quota before a mass roll-out.
+
+## API
+
+Using Memfault’s REST API the devices belonging to a project can be listed
+together with their aspect resolutions. See the
+[API docs](https://api-docs.memfault.com/#2634f3fb-d137-42a8-b093-e05d420b8932)
+for more information.
```
If the API cannot be used to assign resolutions today, we should call this out, IMHO. Something like this maybe?

> There is currently no API to set the Fleet Sampling resolution programmatically. Instead, configure Device Attributes via API and perform a bulk assignment searching for matching devices via the frontend as described above.

HBehrens commented on docs/platform/fleet-sampling.mdx:
```diff
@@ -0,0 +1,198 @@
+[Devices page](https://app.memfault.com/organizations/-/projects/-/devices)
+along with their software versions and last check-in times.
+
+## Managing aspect resolutions of your devices
```
Second thought: Maybe just leave as-is so we don't need to rewrite the sentences that refer to the complementary paragraphs – this way, we should be able to ship this more quickly.

HBehrens reviewed (commented):
Looks good in general! I added two top-level TODOs to the PR description and left more inline comments and suggestions.

can-memfault commented:
> Looks good in general! I added two top-level TODOs to the PR description and left more inline comments and suggestions.

@HBehrens I addressed almost all of your comments, can you take another look once you have time? Thanks!
